### PR TITLE
LU factorization: add allowsingular keyword argument

### LINKED ISSUE #535: [LinearAlgebra] LU-factorization failure (0.7.0-alpha.0)
```julia
julia> A = [1 0 0; .5 1 0; 1 2 1] * [2 0 0; 0 2.5 0; 0 0 0] * [1 -.5 1.5; 0 1 -.2; 0 0 1]
3×3 Array{Float64,2}:
 2.0  -1.0  3.0
 1.0   2.0  1.0
 2.0   4.0  2.0

julia> lu(A)
Failed factorization of type LU{Float64,Array{Float64,2}}

julia> VERSION
v"0.7.0-alpha.0"
```

However Julia 0.4.7 can factorize it

```julia
julia> A = [1 0 0; .5 1 0; 1 2 1] * [2 0 0; 0 2.5 0; 0 0 0] * [1 -.5 1.5; 0 1 -.2; 0 0 1]
3x3 Array{Float64,2}:
 2.0  -1.0  3.0
 1.0   2.0  1.0
 2.0   4.0  2.0

julia> lu(A)
(
3x3 Array{Float64,2}:
 1.0  0.0  0.0
 1.0  1.0  0.0
 0.5  0.5  1.0,

3x3 Array{Float64,2}:
 2.0  -1.0   3.0
 0.0   5.0  -1.0
 0.0   0.0   0.0,

Int32[1,3,2])

julia> VERSION
v"0.4.7"
```

So does octave.

### LINKED ISSUE #27657: [LinearAlgebra] LU-factorization failure (0.7.0-alpha.0)
```julia
julia> A = [1 0 0; .5 1 0; 1 2 1] * [2 0 0; 0 2.5 0; 0 0 0] * [1 -.5 1.5; 0 1 -.2; 0 0 1]
3×3 Array{Float64,2}:
 2.0  -1.0  3.0
 1.0   2.0  1.0
 2.0   4.0  2.0

julia> lu(A)
Failed factorization of type LU{Float64,Array{Float64,2}}

julia> VERSION
v"0.7.0-alpha.0"
```

However Julia 0.4.7 can factorize it

```julia
julia> A = [1 0 0; .5 1 0; 1 2 1] * [2 0 0; 0 2.5 0; 0 0 0] * [1 -.5 1.5; 0 1 -.2; 0 0 1]
3x3 Array{Float64,2}:
 2.0  -1.0  3.0
 1.0   2.0  1.0
 2.0   4.0  2.0

julia> lu(A)
(
3x3 Array{Float64,2}:
 1.0  0.0  0.0
 1.0  1.0  0.0
 0.5  0.5  1.0,

3x3 Array{Float64,2}:
 2.0  -1.0   3.0
 0.0   5.0  -1.0
 0.0   0.0   0.0,

Int32[1,3,2])

julia> VERSION
v"0.4.7"
```

So does octave.

### PR DESCRIPTION
**Pull request by sijow**
Fixes JuliaLang/LinearAlgebra.jl#535.

Currently the `check` argument of `lu` conflates failed factorization (invalid factors) and valid factorizations that have a rank-deficient factor, and the `show` value for the factorization is misleading in the latter case.

This PR adds an `allowsingular` keyword argument to `lu` and the corresponding `issuccess` method to treat valid but rank-deficient factors as success. It also changes `show` to show the factors as usual in such cases, but with a "rank-deficient" note.

The invalid vs rank-defficient factors are distinguished using negative `info` values in the rank-deficient case, as proposed in JuliaLang/LinearAlgebra.jl#535.

### CODE REVIEW

**Comment by mikmoore** [comment]
It seems more Julian to always allow singular and to only to throw if one attempts to use it in an inverse context. But it looks like it would be breaking to make that default, because people who previously depended on a `try`/`catch` or checking `issuccess` to detect and handle a problem would no longer be doing that. Bummer.

Very nice to have this as an option, at least.

File: stdlib/LinearAlgebra/src/lu.jl [hunk 1]
```diff
@@ -76,24 +77,33 @@ Base.iterate(S::LU, ::Val{:done}) = nothing
...
+    lu!(A.data, pivot; check, allowsingular)
 end
 # for backward compatibility
 # TODO: remove towards Julia v2
 @deprecate lu!(A::Union{StridedMatrix,HermOrSym,Tridiagonal}, ::Val{true}; check::Bool = true) lu!(A, RowMaximum(); check=check)
 @deprecate lu!(A::Union{StridedMatrix,HermOrSym,Tridiagonal}, ::Val{false}; check::Bool = true) lu!(A, NoPivot(); check=check)
 
 """
-    lu!(A, pivot = RowMaximum(); check = true) -> LU
+    lu!(A, pivot = RowMaximum(); check = true, allowsingular = false) -> LU
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/lu.jl:106]: Do we need a compat entry for this docstring as well? Doesn't hurt, perhaps?

File: stdlib/LinearAlgebra/src/lu.jl [hunk 2]
```diff
@@ -269,6 +289,9 @@ The relationship between `F` and `A` is
 | [`logabsdet`](@ref)              | ✓    | ✓                      |
 | [`size`](@ref)                   | ✓    | ✓                      |
 
+!!! compat "Julia 1.11"
+    The `allowsingular=` argument was added in Julia 1.11.
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/lu.jl:293]:
```suggestion
    The `allowsingular` keyword argument was added in Julia 1.11.
```

File: stdlib/LinearAlgebra/src/lu.jl [hunk 3]
```diff
@@ -357,17 +391,47 @@ end
...
+
+"""
+    issuccess(F::LU; allowsingular = false)
+
+Test that the LU factorization of a matrix succeeded. By default a
+factorization that produces a valid but rank-deficient U factor is considered a
+failure. This can be changed by passing `allowsingular = true`.
+
+!!! compat "Julia 1.11"
+    The `allowsingular=` argument was added in Julia 1.11.
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/lu.jl:403]:
```suggestion
    The `allowsingular` keyword argument was added in Julia 1.11.
```

File: stdlib/LinearAlgebra/src/lu.jl (hunk 1 above)
Comment (knuesel) [reply stdlib/LinearAlgebra/src/lu.jl:106]: Absolutely, thanks!

**Comment by knuesel** [comment]
Thanks for the review @dkarrasch 

I agree with @mikmoore that it would be better to always allow singular and let the code produce an error later if it's wrongly used, but thought that would be too breaking (I'm happy to make another PR with that behavior if there's agreement that it would be an acceptable change).

**Comment by ViralBShah** [comment]
Is it ok to merge this PR?

**Comment by knuesel** [comment]
I think so (no opposition, and the alternative would be too breaking).

File: stdlib/LinearAlgebra/src/lu.jl [hunk 4]
```diff
@@ -215,14 +233,18 @@ lupivottype(::Type{T}) where {T} = RowMaximum()
...
-    lu(A, pivot = RowMaximum(); check = true) -> F::LU
+    lu(A, pivot = RowMaximum(); check = true, allowsingular = false) -> F::LU
 
 Compute the LU factorization of `A`.
 
 When `check = true`, an error is thrown if the decomposition fails.
 When `check = false`, responsibility for checking the decomposition's
 validity (via [`issuccess`](@ref)) lies with the user.
 
+By default with `check = true` an error is also thrown when the decomposition
```
Comment (jishnub) [inline stdlib/LinearAlgebra/src/lu.jl:244]: It'll be good to have an example demonstrating the use of `allowsingular = true`.

**Comment by jishnub** [comment]
Looks good to me

File: stdlib/LinearAlgebra/src/lu.jl (hunk 4 above)
Comment (knuesel) [reply stdlib/LinearAlgebra/src/lu.jl]: There's the following a the bottom of the `lu` docstring:

```
julia> lu([1 2; 1 2], allowsingular = true)
LU{Float64, Matrix{Float64}, Vector{Int64}}
L factor:
2×2 Matrix{Float64}:
 1.0  0.0
 1.0  1.0
U factor (rank-deficient):
2×2 Matrix{Float64}:
 1.0  2.0
 0.0  0.0
```

Do you mean something else?

Comment (jishnub) [reply stdlib/LinearAlgebra/src/lu.jl]: Oh I had missed that. This looks fine.

**Comment by knuesel** [comment]
Thanks @jishnub . I've just rebased and added a NEWS entry.

**Review by dkarrasch — APPROVED** [review]
Up to the typo, LGTM.

**Comment by rayegun** [comment]
We should have a non-throwing equivalent to the singularity checks correct? To allow exception free checking of factorization singularity without manually checking the info field etc.

**Comment by knuesel** [comment]
You can do `!issuccess(F) && issuccess(F, allowsingular=true)`. I agree the ergonomics are not great but I'm not sure what would be a good solution. An `issingular` function has been [proposed but met with resistance](https://github.com/JuliaLang/julia/issues/27657#issuecomment-405218613). Also if we add something like `isrankdeficient` people will expect it to work for any matrix, but I've seen that [frown upon](https://discourse.julialang.org/t/checking-for-singularity-of-matrix/54746/4).

**Comment by rayegun** [comment]
Could we add a nothrow kwarg to the `check*` functions perhaps?

**Comment by knuesel** [comment]
I think these are private functions meant to simplify control flow in the factorization code, for example a more accurate name for `checknonsingular(info)` would be `throwifsingular(info)`

I'm still not sure a new function is needed: as I understand the `info` field is meant to tell if the algorithm succeeded. For this purpose `issuccess` should be good enough. If a user really cares about the rank of `U` they should probably call `rank`, which works with a known and configurable tolerance...

Maybe it's best to wait for someone to complain with an actual use case?

**Pull request merged**